api/src/pcapi/routes/public/individual_offers/v1/products.py

```diff
@@ -194,6 +196,82 @@ def get_event_titelive_music_types() -> serialization.GetMusicTypesResponse:
     )
 
 
+class CreateProductError(Exception):
+    msg = "can't create this offer"
+
+
+class CreateProductDBError(CreateProductError):
+    msg = "internal error, can't create this offer"
+
+
+class ExistingVenueWithIdAtProviderError(CreateProductDBError):
+    msg = "idAtProvider already exists, can't create this offer"
+
+
+class CreateStockError(CreateProductError):
+    pass
+
+
+class CreateStockDBError(CreateStockError):
+    pass
+
+
+def _create_product(venue: offerers_models.Venue, body: serialization.ProductOfferCreation) -> offers_models.Offer:
+    try:
+        created_product = offers_api.create_offer(
+            audio_disability_compliant=body.accessibility.audio_disability_compliant,
+            booking_contact=body.booking_contact,
+            booking_email=body.booking_email,
+            description=body.description,
+            external_ticket_office_url=body.external_ticket_office_url,
+            extra_data=serialization.deserialize_extra_data(body.category_related_fields),
+            is_duo=body.is_duo,
+            mental_disability_compliant=body.accessibility.mental_disability_compliant,
+            motor_disability_compliant=body.accessibility.motor_disability_compliant,
+            name=body.name,
+            provider=current_api_key.provider,
+            subcategory_id=body.category_related_fields.subcategory_id,
+            url=body.location.url if isinstance(body.location, serialization.DigitalLocation) else None,
+            venue=venue,
+            visual_disability_compliant=body.accessibility.visual_disability_compliant,
+            withdrawal_details=body.withdrawal_details,
+            id_at_provider=body.id_at_provider,
+        )
+
+        # To create stocks or publishing the offer we need to flush
+        # the session to get the offer id
+        db.session.flush()
+    except sqla_exc.IntegrityError as error:
+        # a unique constraint violation can only mean that the venueId/idAtProvider
+        # already exists. Other error are unlikely, but we still need to manage them.
+        if error.orig.diag.table_name == offers_models.Offer.__tablename__:
+            if error.orig.pgcode == UNIQUE_VIOLATION:
+                raise ExistingVenueWithIdAtProviderError() from error
+        raise CreateProductDBError() from error
```

Comment:
Pour rendre plus simple la lecture de ce bout de code, le mieux c'est de faire un seul if et de faire une variable très claire pour la condition qui est checkée.
```suggestion
        # a unique constraint violation can only mean that the venueId/idAtProvider 
        # already exists
        unique_id_at_provider_venue_id_is_violated = error.orig.diag.table_name == offers_models.Offer.__tablename__ and error.orig.pgcode == UNIQUE_VIOLATION
        if unique_id_at_provider_venue_id_is_violated:
             raise ExistingVenueWithIdAtProviderError() from error
        # Other error are unlikely, but we still need to manage them.
        raise CreateProductDBError() from error
```

Reply:
Deuxième remarque : je crois qu'il est possible de simplifier le check en regardant le nom de la règle violée qui a un nom unique (`unique_idAtProvider_venueId`)